text/3804-cfg-alias.md

```diff
@@ -0,0 +1,165 @@
+group around, which is error-prone and noisy.
+
+Adding aliases to config predicates reduces the amount of code that needs to be
+duplicated, and giving it a name provides an easy way to show what a group of
+configuration is intended to represent.
+
+Something to this effect can be done using build scripts. This requires reading
+various Cargo environment variables and potentially doing string manipulation
+(for splitting target features), so it is often inconvenient enough to not be
+worth doing. Allowing aliases to be defined within the crate and with the same
+syntax as the `cfg` itself makes this much easier.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+There is a new crate-level attribute that takes a name and a `cfg` predicate:
+
+```rust
+#![cfg_alias(some_alias, predicate)]
+```
+
+`predicate` can be anything that usually works within `#[cfg(...)]`, including
+`all`, `any`, and `not`.
+
+Once an alias is defined, `name` can be used as if it had been passed via
+`--cfg`:
+
+```rust
+#[cfg(some_alias)]
+struct Foo { /* ... */ }
+
+#[cfg(not(some_alias))]
+struct Foo { /* ... */ }
+
+#[cfg(all(some_alias, target_os = "linux"))]
+fn bar() { /* ... */ }
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The new crate-level attribute is introduced:
+
+```text
+CfgAliasAttribute:
+    cfg_alias(IDENTIFIER, ConfigurationPredicate)
+```
+
+The identifier is added to the `cfg` namespace. It must not conflict with any
+builtin configuration names, or with those passed via `--cfg`.
+
+Once defined, the alias can be used as a regular predicate.
+
+The alias is only usable after it has been defined. For example, the following
+will emit an unknown configuration lint:
+
+```rust
+#![cfg_attr(some_alias, some_attribute)]
+// warning: unexpected_cfgs
+//
+// The lint could mention that `some_alias` was found in the
+// crate but is not available here.
+
+#![cfg_alias(some_alias, true)]
+```
+
+RFC Question:
+
+Two ways to implement this are with (1) near-literal substitution, or (2)
+checking whether the alias should be set or not at the time it is defined. Is
+there any user-visible behavior that would make us need to specify one or the
+other?
+
+If we go with the first option, we should limit to a single expansion to avoid
+recursing (as is done for `#define` in C).
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- This does not support more general attribute aliases, such as
+  `#![alias(foo = derive(Clone, Copy, Debug, Default)`. This seems better suited
+  for something like `declarative_attribute_macros` in [RFC3697].
+
+[RFC3697]: https://github.com/rust-lang/rfcs/pull/3697
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The syntax `cfg_alias(name, predicate)` was chosen for similarity with
+  `cfg_attr(predicate, attributes)`. Alternatives include:
+  - `cfg_alias(name = predicate)`
```

Comment:
I prefer the `=` form. `cfg_attr` doesn't seem like a compelling precedent when the predicate goes first with `cfg_attr` and second with `cfg_alias`.

I also happen to think the `cfg_attr` syntax is hard to read, we should avoid replicating that :)

Reply:
`cfg_attr` is definitely not very nice to read or write (though, I don't really know what would have been better here). Reading through it more, I am also starting to prefer `=` but one concern is that `=` in config is used for something like equality rather than assignment. Should there be any concern about that here?

I know that also came up in discussion at https://github.com/rust-lang/rfcs/pull/3796

Reply:
I have updated the syntax to use `=`

Reply:
One other downside: with `=`, it looks a bit weird when aliasing a single `=` option

```rust
#![cfg_alias(alias = target_os = "linux")]
```

Reply:
Ah, that is a negative. I'll retract my preference then. I'm not crazy about the comma but it's closest to what we have..

Reply:
We could choose to require parens there.

Reply:
Using a comma for this is indeed the most consistent with what we do today.  I'd probably suggest to switch back to that and save doing something with `=` for the day when we work out a plan to stop using `=` for subset operations in attribute position.